File: backend/app/api/topology.py
"""Thin API wrappers for the topology and asset drill-down surface."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Header, HTTPException, Query

from app import schemas
from app.core.db import get_db
from app.repositories.topology import has_topology_schema as _repo_has_topology_schema
from app.services.brinqa_detail import BrinqaAuthContext, asset_detail_service
from app.services.topology_view import (
    get_application_detail as _get_application_detail,
    get_asset_detail as _get_asset_detail,
    get_asset_enrichment as _get_asset_enrichment,
    get_asset_findings as _get_asset_findings,
    get_asset_findings_analytics as _get_asset_findings_analytics,
    get_assets as _get_assets,
    get_assets_analytics as _get_assets_analytics,
    get_business_service_analytics as _get_business_service_analytics,
    get_business_service_detail as _get_business_service_detail,
    get_business_unit_detail as _get_business_unit_detail,
    get_business_units as _get_business_units,
)

router = APIRouter(tags=["topology"])


def _has_topology_schema(db):
    return _repo_has_topology_schema(db)


@router.get("/topology/business-units", response_model=list[schemas.BusinessUnitSummary])
def get_business_units(db=Depends(get_db)):
    if not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "Normalized topology tables are not initialized. "
                "Apply docs/backend/topology-seed/topology-expansion.sql before using "
                "business-unit topology routes."
            ),
        )
    return _get_business_units(db)


@router.get("/topology/business-units/{business_unit_slug}", response_model=schemas.BusinessUnitDetail)
def get_business_unit_detail(business_unit_slug: str, db=Depends(get_db)):
    if not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "Normalized topology tables are not initialized. "
                "Apply docs/backend/topology-seed/topology-expansion.sql before using "
                "business-unit topology routes."
            ),
        )
    return _get_business_unit_detail(business_unit_slug, db)


@router.get(
    "/topology/business-units/{business_unit_slug}/business-services/{business_service_slug}",
    response_model=schemas.BusinessServiceDetail,
)
def get_business_service_detail(
    business_unit_slug: str,
    business_service_slug: str,
    db=Depends(get_db),
):
    if not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "Normalized topology tables are not initialized. "
                "Apply docs/backend/topology-seed/topology-expansion.sql before using "
                "business-unit topology routes."
            ),
        )
    return _get_business_service_detail(business_unit_slug, business_service_slug, db)


@router.get(
    "/topology/business-units/{business_unit_slug}/business-services/{business_service_slug}/analytics",
    response_model=schemas.BusinessServiceAnalytics,
)
def get_business_service_analytics(
    business_unit_slug: str,
    business_service_slug: str,
    db=Depends(get_db),
):
    if not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "Normalized topology tables are not initialized. "
                "Apply docs/backend/topology-seed/topology-expansion.sql before using "
                "business-unit topology routes."
            ),
        )
    return _get_business_service_analytics(business_unit_slug, business_service_slug, db)


@router.get(
    "/topology/business-units/{business_unit_slug}/business-services/{business_service_slug}/applications/{application_slug}",
    response_model=schemas.ApplicationDetail,
)
def get_application_detail(
    business_unit_slug: str,
    business_service_slug: str,
    application_slug: str,
    db=Depends(get_db),
):
    if not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "Normalized topology tables are not initialized. "
                "Apply docs/backend/topology-seed/topology-expansion.sql before using "
                "business-unit topology routes."
            ),
        )
    return _get_application_detail(business_unit_slug, business_service_slug, application_slug, db)


@router.get("/assets", response_model=schemas.PaginatedAssets)
def get_assets(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    business_unit: str | None = Query(None),
    business_service: str | None = Query(None),
    application: str | None = Query(None),
    status: str | None = Query(None),
    environment: str | None = Query(None),
    compliance: str | None = Query(None),
    search: str | None = Query(None),
    direct_only: bool = Query(False),
    sort_by: str = Query("name"),
    sort_order: str = Query("asc"),
    db=Depends(get_db),
):
    if business_unit is not None and not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "The business_unit filter requires the normalized topology schema. "
                "Apply docs/backend/topology-seed/topology-expansion.sql first."
            ),
        )
    return _get_assets(
        page=page,
        page_size=page_size,
        business_unit=business_unit,
        business_service=business_service,
        application=application,
        status=status,
        environment=environment,
        compliance=compliance,
        search=search,
        direct_only=direct_only,
        sort_by=sort_by,
        sort_order=sort_order,
        db=db,
    )


@router.get("/assets/analytics", response_model=schemas.AssetAnalyticsResponse)
def get_assets_analytics(
    business_unit: str | None = Query(None),
    business_service: str | None = Query(None),
    application: str | None = Query(None),
    status: str | None = Query(None),
    environment: str | None = Query(None),
    compliance: str | None = Query(None),
    search: str | None = Query(None),
    direct_only: bool = Query(False),
    db=Depends(get_db),
):
    if business_unit is not None and not _has_topology_schema(db):
        raise HTTPException(
            status_code=503,
            detail=(
                "The business_unit filter requires the normalized topology schema. "
                "Apply docs/backend/topology-seed/topology-expansion.sql first."
            ),
        )
    return _get_assets_analytics(
        business_unit=business_unit,
        business_service=business_service,
        application=application,
        status=status,
        environment=environment,
        compliance=compliance,
        search=search,
        direct_only=direct_only,
        db=db,
    )


@router.get("/assets/{asset_id}", response_model=schemas.AssetDetail)
def get_asset_detail(asset_id: str, db=Depends(get_db)):
    return _get_asset_detail(asset_id, db)


@router.get("/assets/{asset_id}/enrichment", response_model=schemas.AssetEnrichment)
def get_asset_enrichment(
    asset_id: str,
    db=Depends(get_db),
    x_brinqa_auth_token: str | None = Header(None, alias="X-Brinqa-Auth-Token"),
    x_brinqa_session_cookie: str | None = Header(None, alias="X-Brinqa-Session-Cookie"),
):
    return _get_asset_enrichment(
        asset_id,
        db,
        x_brinqa_auth_token=x_brinqa_auth_token,
        x_brinqa_session_cookie=x_brinqa_session_cookie,
    )


@router.get("/assets/{asset_id}/findings", response_model=schemas.AssetFindingsPage)
def get_asset_findings(
    asset_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=200),
    sort_by: str = Query("risk_score"),
    sort_order: str = Query("desc"),
    risk_band: str | None = Query(None),
    kev_only: bool = Query(False),
    source: str | None = Query(None),
    search: str | None = Query(None),
    db=Depends(get_db),
):
    return _get_asset_findings(
        asset_id,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
        risk_band=risk_band,
        kev_only=kev_only,
        source=source,
        search=search,
        db=db,
    )


@router.get(
    "/assets/{asset_id}/findings/analytics",
    response_model=schemas.AssetFindingsAnalyticsResponse,
)
def get_asset_findings_analytics(
    asset_id: str,
    risk_band: str | None = Query(None),
    kev_only: bool = Query(False),
    source: str | None = Query(None),
    search: str | None = Query(None),
    db=Depends(get_db),
):
    return _get_asset_findings_analytics(
        asset_id,
        risk_band=risk_band,
        kev_only=kev_only,
        source=source,
        search=search,
        db=db,
    )

File: frontend/src/api/topology.ts
import { buildApiUrl, buildBrinqaEnrichmentRequestInit, parseJsonOrThrow } from "./client";
import type {
  AssetListSortBy,
  AssetFindingsAnalyticsResponse,
  AssetAnalyticsResponse,
  ApplicationDetail,
  AssetDetail,
  AssetEnrichment,
  AssetFindingsPage,
  BusinessServiceDetail,
  BusinessUnitDetail,
  BusinessUnitSummary,
  FindingsSortBy,
  PaginatedAssets,
  SortOrder,
} from "./types";

export async function getBusinessUnits(): Promise<BusinessUnitSummary[]> {
  const res = await fetch(buildApiUrl("/topology/business-units"));
  return parseJsonOrThrow(
    res,
    `Failed to fetch business units: ${res.status} ${res.statusText}`
  );
}

export async function getBusinessUnitDetail(
  businessUnitSlug: string
): Promise<BusinessUnitDetail> {
  const res = await fetch(buildApiUrl(`/topology/business-units/${businessUnitSlug}`));
  return parseJsonOrThrow(
    res,
    `Failed to fetch business unit detail: ${res.status} ${res.statusText}`
  );
}

export async function getBusinessServiceDetail(
  businessUnitSlug: string,
  businessServiceSlug: string
): Promise<BusinessServiceDetail> {
  const res = await fetch(
    buildApiUrl(
      `/topology/business-units/${businessUnitSlug}/business-services/${businessServiceSlug}`
    )
  );
  return parseJsonOrThrow(
    res,
    `Failed to fetch business service detail: ${res.status} ${res.statusText}`
  );
}

export async function getApplicationDetail(
  businessUnitSlug: string,
  businessServiceSlug: string,
  applicationSlug: string
): Promise<ApplicationDetail> {
  const res = await fetch(
    buildApiUrl(
      `/topology/business-units/${businessUnitSlug}/business-services/${businessServiceSlug}/applications/${applicationSlug}`
    )
  );
  return parseJsonOrThrow(
    res,
    `Failed to fetch application detail: ${res.status} ${res.statusText}`
  );
}

export async function getAssetDetail(assetId: string): Promise<AssetDetail> {
  const res = await fetch(buildApiUrl(`/assets/${assetId}`));
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset detail: ${res.status} ${res.statusText}`
  );
}

export async function getAssetEnrichment(assetId: string): Promise<AssetEnrichment> {
  const res = await fetch(
    buildApiUrl(`/assets/${assetId}/enrichment`),
    buildBrinqaEnrichmentRequestInit()
  );
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset enrichment: ${res.status} ${res.statusText}`
  );
}

export async function getAssetFindings(assetId: string): Promise<AssetFindingsPage> {
  const res = await fetch(buildApiUrl(`/assets/${assetId}/findings`));
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset findings: ${res.status} ${res.statusText}`
  );
}

export async function getAssetFindingsAnalytics(
  assetId: string,
  {
    riskBand,
    kevOnly,
    source,
    search,
  }: {
    riskBand: "All" | "Critical" | "High" | "Medium" | "Low";
    kevOnly: boolean;
    source?: string | null;
    search?: string | null;
  }
): Promise<AssetFindingsAnalyticsResponse> {
  const params = new URLSearchParams();
  if (riskBand !== "All") {
    params.set("risk_band", riskBand);
  }
  if (kevOnly) {
    params.set("kev_only", "true");
  }
  if (source && source !== "All") {
    params.set("source", source);
  }
  if (search) {
    params.set("search", search);
  }
  const res = await fetch(buildApiUrl(`/assets/${assetId}/findings/analytics`, params));
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset findings analytics: ${res.status} ${res.statusText}`
  );
}

export async function getAssetFindingsPage(
  assetId: string,
  {
    page,
    pageSize,
    sortBy,
    sortOrder,
    riskBand,
    kevOnly,
    source,
    search,
  }: {
    page: number;
    pageSize: number;
    sortBy: FindingsSortBy;
    sortOrder: SortOrder;
    riskBand: "All" | "Critical" | "High" | "Medium" | "Low";
    kevOnly: boolean;
    source?: string | null;
    search?: string | null;
  }
): Promise<AssetFindingsPage> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
    sort_by: sortBy,
    sort_order: sortOrder,
  });
  if (riskBand !== "All") {
    params.set("risk_band", riskBand);
  }
  if (kevOnly) {
    params.set("kev_only", "true");
  }
  if (source && source !== "All") {
    params.set("source", source);
  }
  if (search) {
    params.set("search", search);
  }
  const res = await fetch(buildApiUrl(`/assets/${assetId}/findings`, params));
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset findings: ${res.status} ${res.statusText}`
  );
}

export async function getAssetsPage({
  page,
  pageSize,
  businessUnit,
  businessService,
  application,
  status,
  environment,
  compliance,
  search,
  directOnly,
  sortBy,
  sortOrder,
}: {
  page: number;
  pageSize: number;
  businessUnit?: string | null;
  businessService?: string | null;
  application?: string | null;
  status?: string | null;
  environment?: string | null;
  compliance?: string | null;
  search?: string | null;
  directOnly?: boolean;
  sortBy: AssetListSortBy;
  sortOrder: SortOrder;
}): Promise<PaginatedAssets> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
    sort_by: sortBy,
    sort_order: sortOrder,
  });
  if (businessUnit) params.set("business_unit", businessUnit);
  if (businessService) params.set("business_service", businessService);
  if (application) params.set("application", application);
  if (status) params.set("status", status);
  if (environment) params.set("environment", environment);
  if (compliance) params.set("compliance", compliance);
  if (search) params.set("search", search);
  if (directOnly) params.set("direct_only", "true");
  const res = await fetch(buildApiUrl("/assets", params));
  return parseJsonOrThrow(
    res,
    `Failed to fetch assets: ${res.status} ${res.statusText}`
  );
}

export async function getAssetsAnalytics({
  businessUnit,
  businessService,
  application,
  status,
  environment,
  compliance,
  search,
  directOnly,
}: {
  businessUnit?: string | null;
  businessService?: string | null;
  application?: string | null;
  status?: string | null;
  environment?: string | null;
  compliance?: string | null;
  search?: string | null;
  directOnly?: boolean;
}): Promise<AssetAnalyticsResponse> {
  const params = new URLSearchParams();
  if (businessUnit) params.set("business_unit", businessUnit);
  if (businessService) params.set("business_service", businessService);
  if (application) params.set("application", application);
  if (status) params.set("status", status);
  if (environment) params.set("environment", environment);
  if (compliance) params.set("compliance", compliance);
  if (search) params.set("search", search);
  if (directOnly) params.set("direct_only", "true");
  const res = await fetch(buildApiUrl("/assets/analytics", params));
  return parseJsonOrThrow(
    res,
    `Failed to fetch asset analytics: ${res.status} ${res.statusText}`
  );
}
